components/feature/privatemode/src/main/java/mozilla/components/feature/privatemode/notification/AbstractPrivateNotificationService.kt

```diff
@@ -0,0 +1,138 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.selector.privateTabs
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.feature.privatemode.R
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.ids.SharedIdsHelper
+import mozilla.components.support.ktx.android.notification.ChannelData
+import mozilla.components.support.ktx.android.notification.ensureNotificationChannelExists
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Manages notifications for private tabs.
+ *
+ * Private tab notifications solve two problems:
+ * 1. They allow users to interact with the browser from outside of the app
+ *    (example: by closing all private tabs).
+ * 2. The notification will keep the process alive, allowing the browser to
+ *    keep private tabs in memory.
+ *
+ * As long as a private tab is open this service will keep its notification alive.
+ */
+abstract class AbstractPrivateNotificationService : Service() {
+
+    private var scope: CoroutineScope? = null
+
+    abstract val store: BrowserStore
+
+    /**
+     * Customizes the private browsing notification.
+     */
+    abstract fun NotificationCompat.Builder.buildNotification()
+
+    /**
+     * Erases all private tabs in reaction to the user tapping the notification.
+     */
+    @CallSuper
+    protected open fun erasePrivateTabs() {
+        store.dispatch(TabListAction.RemoveAllPrivateTabsAction)
+    }
+
+    /**
+     * Create the private browsing notification and
+     * add a listener to stop the service once all private tabs are closed.
+     *
+     * The service should be started only if private tabs are open.
+     */
+    @ExperimentalCoroutinesApi
+    final override fun onCreate() {
+        val id = SharedIdsHelper.getIdForTag(this, NOTIFICATION_TAG)
+        val channelId = ensureNotificationChannelExists(this, NOTIFICATION_CHANNEL, onSetupChannel = {
+            if (SDK_INT >= Build.VERSION_CODES.O) {
+                enableLights(false)
+                enableVibration(false)
+                setShowBadge(false)
+            }
+        })
+
+        val notification = NotificationCompat.Builder(this, channelId)
+            .setOngoing(true)
+            .setVisibility(VISIBILITY_SECRET)
+            .setShowWhen(false)
+            .setLocalOnly(true)
+            .setContentIntent(Intent(ACTION_ERASE).let {
+                it.setClass(this, this::class.java)
+                PendingIntent.getService(this, 0, it, FLAG_ONE_SHOT)
+            })
+            .apply { buildNotification() }
+            .build()
+
+        startForeground(id, notification)
+
+        scope = store.flowScoped { flow ->
```

Comment:
Just found a memory leak in the Fenix version of this: https://github.com/mozilla-mobile/fenix/pull/10789#discussion_r445678387

This solution here would fix this! So, 👍 .